common/lib/client/resource.ts

```diff
@@ -0,0 +1,173 @@
+			callback(wrappedErr, response, wrappedHeaders, true, wrappedStatusCode);
+			return;
+		}
+
+		callback(err, response, wrappedHeaders, true, wrappedStatusCode);
+	};
+}
+
+function paramString(params: Record<string, any>) {
+	const paramPairs = [];
+	if (params) {
+		for (const needle in params) {
+			paramPairs.push(needle + '=' + params[needle]);
+		}
+	}
+	return paramPairs.join('&');
+}
+
+function urlFromPathAndParams(path: string, params: Record<string, any>) {
+	return path + (params ? '?' : '') + paramString(params);
+}
+
+function logResponseHandler(callback: Function, method: HttpMethods, path: string, params: Record<string, string>) {
+	return function(err: Error, body: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number) {
+		if (err) {
+			Logger.logAction(Logger.LOG_MICRO, 'Resource.' + method + '()', 'Received Error; ' + urlFromPathAndParams(path, params) + '; Error: ' + Utils.inspectError(err));
+		} else {
+			Logger.logAction(Logger.LOG_MICRO, 'Resource.' + method + '()',
+				'Received; ' + urlFromPathAndParams(path, params) + '; Headers: ' + paramString(headers) + '; StatusCode: ' + statusCode + '; Body: ' + (BufferUtils.isBuffer(body) ? body.toString() : body));
+		}
+		if (callback) { callback(err, body, headers, unpacked, statusCode); }
+	}
+}
+
+class Resource {
+	static get(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Get, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static delete(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Delete, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static post(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Post, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static patch(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Patch, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static put(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Put, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static do(method: HttpMethods, rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		if (Logger.shouldLog(Logger.LOG_MICRO)) {
```

Comment:
This if may not be necessary

Reply:
This is for performance. If we call `Logger.logAction` then even if the `logLevel` is above `LOG_MICRO` it will still serialise the path and params into a string in order to call the function. This if statement makes it so that the serialisation only happens if it needs to.

Reply:
Resolving this conversation on @ikbalkaya's behalf.